hack/verify-terraform.sh

```diff
@@ -20,46 +20,77 @@ set -o pipefail
 REPO_ROOT=$(cd "$(dirname "${BASH_SOURCE[0]}")/.." && pwd -P)
 
 function usage() {
-  echo >&2 "Usage: $0 <PATH>"
-  exit 1
+    echo >&2 "Usage: $0 [path...]"
+    exit 1
+}
+
+function ensure_dependencies() {
+    if [ -z "$(which curl)" ]; then
+        echo "Please install curl"
+        exit 1
+    fi
+
+    if [ -z "$(which tfswitch)" ]; then
+        echo "Please install tfswitch"
+        exit 1
+    fi
 }
 
 function check_terraform() {
-  if [ $# != 1 ]; then
-    echo "check_terraform(path) requires 1 argument" >&2
-    exit 1
-  fi
+  local path="${1}"
 
-  if [ ! -d "$1" ]; then
-    echo "path not found"
-    exit 1
+  if [ ! -d "${path}" ]; then
+      echo "check_terraform: path '${path}' is not a directory (or does not exist)"
+      return 1
   fi
 
-  cd "$1"
-  echo "Installing Terraform"
+  pushd "${path}" >/dev/null
+
+  echo "# Installing terraform for path: ${path} ..."
   tfswitch
-  echo "Running terraform validate"
+
+  echo "# Running terraform validate for path: ${path} ..."
   terraform init -backend=false
   terraform validate
+
+  popd >/dev/null
 }
 
+function main() {
+    local paths=("$@")
+    local failures=()
 
-if [ $# != 1 ]; then
-    usage
-    exit 1
-fi
-
-if [ -z "$(which curl)" ]; then
-  echo "Please install curl"
-  exit 1
-fi
-
-if [ -z "$(which tfswitch)" ]; then
-  echo "Installing tfswitch locally"
-  curl -L https://raw.githubusercontent.com/warrensbox/terraform-switcher/release/install.sh | bash
-else
-tfswitch --version
-fi
-
-cd "$REPO_ROOT"
-check_terraform "$1"
+    ensure_dependencies
+
+    # if no paths specified, default to all paths that contain *.tf
+    pushd "$REPO_ROOT"
+    if [ "${#paths[@]}" == 0 ]; then
+        mapfile -t paths < <(
+          find . -name '*.tf' -print0 | xargs -0 -n1 dirname | sort | uniq
```

Comment:
The results include the terraform modules(/infra/gcp/clusters/modules). Do we need to validate them knowing `check_terraform ${path}` already does this part ?

Reply:
I'll update to exclude them.  My pedantic concern was that would allow us to checkin dead / broken code, since we have nothing to verify/enforce that all modules are used at least once.  But I think you're right, it's probably overkill

Reply:
That saves some runtime, but it still feels not quite ready to run by default all the time to me.  Most PR's in this repo don't touch terraform.  But we could probably go PR test-infra to have a pull-k8sio-verify-terraform job that triggers for `.tf$` instead of the per-cluster jobs we have now.   That's what motivated this in the first place, I realized we were missing some of those jobs for the new projects of late.